server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.util.concurrent;
+
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.TestThreadPool;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+
+public class PrioritizedThrottledTaskRunnerTests extends ESTestCase {
+    private ThreadPool threadPool;
+    private Executor executor;
+
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        threadPool = new TestThreadPool("test");
+        executor = threadPool.executor(ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+        TestThreadPool.terminate(threadPool, 30, TimeUnit.SECONDS);
+    }
+
+    static class TestTask implements Comparable<TestTask>, Runnable {
+
+        private final Runnable runnable;
+        private final int priority;
+
+        TestTask(Runnable runnable, int priority) {
+            this.runnable = runnable;
+            this.priority = priority;
+        }
+
+        @Override
+        public int compareTo(TestTask o) {
+            return priority - o.priority;
+        }
+
+        @Override
+        public void run() {
+            runnable.run();
+        }
+    }
+
+    public void testMultiThreadedEnqueue() throws Exception {
+        final int maxTasks = randomIntBetween(1, threadPool.info(ThreadPool.Names.GENERIC).getMax());
+        PrioritizedThrottledTaskRunner<TestTask> taskRunner = new PrioritizedThrottledTaskRunner<>("test", maxTasks, executor);
+        final int enqueued = randomIntBetween(2 * maxTasks, 10 * maxTasks);
+        AtomicInteger executed = new AtomicInteger();
+        CountDownLatch threadBlocker = new CountDownLatch(enqueued);
+        for (int i = 0; i < enqueued; i++) {
+            new Thread(() -> {
+                try {
+                    threadBlocker.countDown();
```

Comment:
NIT: could use `CyclicBarrier` here instead which seems to be the correct primitive?

Reply:
We could. But since we need only a one-time barrier, the latch is enough I think. I've seen it used like that in several places in the code.

Reply:
I'd agree with Armin here, `CyclicBarrier` is good for a rendezvous interaction. And saves a line of code too.

But fine to leave as is...